From eb85075070920f9711b851bd03a3aac85c092fc2 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Przemys=C5=82aw=20Szczepaniak?= <pszczepaniak@google.com>
Date: Mon, 13 Mar 2023 14:38:28 +0000
Subject: [PATCH] Fix OOB Read in setOperandValue

Bug: 269456018
Test: Run the POC
(cherry picked from https://android-review.googlesource.com/q/commit:c45bdb6ac47bf8cf2853144e82910f43f2f0b1e9)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:b108bc94105b1a12f0fcf21bf7e13ccdee6e4efa)
Merged-In: I7325d56a380f05753356875623a2b5eaba3ca578
Change-Id: I7325d56a380f05753356875623a2b5eaba3ca578
---
 shim_and_sl/ShimConverter.cpp | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/shim_and_sl/ShimConverter.cpp b/shim_and_sl/ShimConverter.cpp
index 398cc9b9..b0f914c3 100644
--- a/shim_and_sl/ShimConverter.cpp
+++ b/shim_and_sl/ShimConverter.cpp
@@ -128,6 +128,12 @@ ANeuralNetworksModel* convertSubgraphFromHAL(
 
         switch (operand.lifetime) {
             case OperandLifeTime::CONSTANT_COPY: {
+                if (operand.location.length + operand.location.offset >
+                    model.operandValues.size()) {
+                    *errorStatus = ErrorStatus::INVALID_ARGUMENT;
+                    return nullptr;
+                }
+
                 if (operand.location.length <=
                     ANEURALNETWORKS_MAX_SIZE_OF_IMMEDIATELY_COPIED_VALUES) {
                     resultModel.setOperandValue(
-- 
2.41.0.255.g8b1d071c50-goog

